proxy.go

```diff
@@ -408,6 +413,37 @@ func requireBasicAuth(authenticator *auth.BasicAuth, handler func(http.ResponseW
 	})
 }
 
+func WithSuitableAuthentication(authenticator *auth.BasicAuth, handler func(http.ResponseWriter, *http.Request)) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if !guestAccessAllowed {
+			//All requests require authentication
+			delegate := requireBasicAuth(authenticator, handler)
+			delegate(w, r)
+		} else {
+			_, _, basicAuthPresent := r.BasicAuth()
+			if !basicAuthPresent {
+				//Run the handler as unauthenticated user
+				confLock.RLock()
+				_, ok := quota[guestUserName]
+				confLock.RUnlock()
+				if !ok {
+					reply(w, errMsg("Guest access is unavailable."), http.StatusUnauthorized)
+				} else {
+					handler(w, r)
+				}
+			} else {
+				//Check that authentication is correct
+				userName := authenticator.CheckAuth(r)
```

Comment:
Take a look at `authenticator.Wrap` method implementation used inside `requireBasicAuth`. It seems that this branch is a duplicate of this code. In that case the entire method can be simplified to something like:
```
func WithSuitableAuthentication(authenticator *auth.BasicAuth, handler func(http.ResponseWriter, *http.Request)) func(http.ResponseWriter, *http.Request) {
        return func(w http.ResponseWriter, r *http.Request) {
                _, _, basicAuthPresent := r.BasicAuth()
                if !basicAuthPresent && guestAccessAllowed {
                        // Guest authentication logic as currently implemented
                        // ...
                } else {
                        requireBasicAuth(authenticator, handler)(w, r)
                }
        }
}
```